src/python/bcc/usdt.py

```diff
@@ -0,0 +1,433 @@
+                        return \
+"""{
+        u64 __temp = ctx->%s + (%d);
+        bpf_probe_read(&%s, sizeof(%s), (void *)__temp);
+}                       """ % (normalized_reg, self.deref_offset,
+                               local_name, local_name)
+
+                # Final case: dereference global, need to find address of global
+                # with the provided name and then potentially add deref_offset
+                # and bpf_probe_read the result. None of this will work with BPF
+                # because we can't just access arbitrary addresses.
+                return "%s = 0;      /* UNSUPPORTED CASE, SEE SOURCE */" % \
+                        local_name
+
+        def get_type(self):
+                result_type = None
+                if self.size == 1:
+                        result_type = "char"
+                elif self.size == 2:
+                        result_type = "short"
+                elif self.size == 4:
+                        result_type = "int"
+                elif self.size == 8:
+                        result_type = "long"
+
+                if result_type is None:
+                        raise ValueError("arguments of size %d are not " +
+                                         "currently supported" % self.size)
+
+                if not self.is_signed:
+                        result_type = "unsigned " + result_type
+
+                return result_type
+
+        def __str__(self):
+                prefix = "%d %s bytes @ " % (self.size,
+                        "  signed" if self.is_signed else "unsigned")
+                if self.constant is not None:
+                        return prefix + "constant %d" % self.constant
+                if self.deref_offset is None:
+                        return prefix + "register " + self.register
+                if self.deref_offset is not None and self.deref_name is None:
+                        return prefix + "%d(%s)" % (self.deref_offset,
+                                                    self.register)
+                return prefix + "%d from %s global" % (self.deref_offset,
+                                                       self.deref_name)
+
+class USDTProbeLocation(object):
+        def __init__(self, address, args):
+                self.address = address
+                self.raw_args = args
+                self.args = []
+                self._parse_args()
+
+        def generate_usdt_assignments(self, prefix="arg"):
+                text = ""
+                for i, arg in enumerate(self.args, 1):
+                        text += (" "*16) + \
+                                arg.generate_assign_to_local(
+                                                "%s%d" % (prefix, i)) + "\n"
+                return text
+
+        def _parse_args(self):
+                for arg in self.raw_args.split():
+                        self._parse_arg(arg.strip())
+
+        def _parse_arg(self, arg):
```

Comment:
Something like 'usdt prerequisites' or user focused doc is needed. Or faq that describes what should be in the binary for this mechanism to work. May be including 'readelf -n|grep magic', so that users can do quick sanity test that stap style usdts are there.